serverless/serverless
# WIP: Add high resource count support

**Note:** This is the first naive iteration. It's not 100% working right now and a better splitting algorithm (e.g. based on resource groups) is needed! Furthermore it's a WIP. The deployment fails because of unresolved refs etc.

## Todos

- [x] Build dependency graph to resolve `Refs` etc.
- [ ] Switch from naive stack splitting to resource grouped splitting
- [ ] DRY out code / fix Todo comments
- [x] Run `npm shrinkwrap` (currently fails on my machine)
- [ ] Print log message when resource count is nearly reached + add link do docs for stack splitting
- [ ] Translate AWS error message ("Resource count is over 200@") into useful error message + add link do docs for stack splitting
- [ ] Use `provider.naming` to centralize resource naming (e.g. nested stack resource logical ids)

---

## What did you implement:

Closes #3411

Add automatic stack splitting when service contains a high number of resources.

## How did you implement it:

Serverless will check the compiled CloudFormation template before deploying to AWS. It will split the stack up into nested stacks if the resource count is above X resources.

After that it will upload the nested stack templates to the S3 Bucket and update the compiled CloudFormation template which is in memory. This makes this whole change encapsulated and backwards compatible.

The user has to specifically opt-in for this feature via the `useStackSplitting: true` config.

## How can we verify it:

Create a service which looks like the following:

```yml
service: service

provider:
  name: aws
  runtime: nodejs4.3
  useStackSplitting: true

functions:
  hello:
    handler: handler.hello
    events:
      - http:
          method: ANY
          path: hello
          integration: LAMBDA
          cors: true
  goodbye:
    handler: handler.goodbye
    events:
      - http:
          method: GET
          path: goodbye
```

Run `serverless deploy --noDeploy` and look into the `.serverless` directory to see all the resources.

Or run `serverless deploy` to deploy the stack.

---

***Is this ready for review?:*** NO
***Is it a breaking change?:*** NO

/cc @brianneisler @eahefnawy @dougmoscrop
